gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java

```diff
@@ -684,6 +743,187 @@ public Boolean call() {
     }
   }
 
+  @Override
+  public Address get(final AddressId addressId, AddressOption... options) {
+    final Map<ComputeRpc.Option, ?> optionsMap = optionMap(options);
+    try {
+      com.google.api.services.compute.model.Address answer =
+          runWithRetries(new Callable<com.google.api.services.compute.model.Address>() {
+            @Override
+            public com.google.api.services.compute.model.Address call() {
+              switch (addressId.type()) {
+                case REGION:
+                  RegionAddressId regionAddressId = (RegionAddressId) addressId;
```

Comment:
> I don't think this is a problem as long as we check and throw if the type is not expected.

In this case you may get a runtime ClassCastException if one provided `AddressId` that returns a known type but is a user supplied implementation.

> How would having interface inheritance per se avoid casting? 

Yes, the main problem is not with casting per se but rather with the expectation that we 
can cast to known (our) implementation but yet nothing enforce that or documents that.
At the very least we could document `AddressId.Type` enum values to say what is the
expected implementation per type value (but I still think it is a design compromise and would favor code duplication over it, if we can make it abstract class instead).

> For create instead we could have two overloaded addressId setters in the AccessInfo builder that only accept the available implementations. 

Yes, I agree. This is better (I also mentioned it above), however I think just this change by itself is not enough.

Another thinking out-loud, if we can't make `AddressId` a class than maybe we can change `RegionAddressId` and `GlobalAddressId` to be public interfaces extending `AddressId` and having our own "internal" implementation of those. When we get an instance of `AddressId`
we could copy the values to our own internal type (if necessary). In this case I will not have
the enum type and rather check instanceof the specific interfaces.

Yet, another option (which I guess is reasonable if the set of difference between sub types is small) is to, as you suggested, have them all in one class (or interface) and use the type enum as a documented behaviour that suggests when some of the fields are populated.